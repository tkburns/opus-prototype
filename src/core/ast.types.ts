import { TokenMap } from './lexer';

export type AST = Program;
export type Node = (
  Program |
  Declaration |
  FuncCall |
  Func |
  Tuple |
  Name |
  Numeral |
  Text
);


export type Program = {
  type: 'program';
  declarations: Declaration[];
  topExpression: Expression;
};

export type Declaration = {
  type: 'declaration';
  name: Name;
  expression: Expression;
};

export type Expression = (
  FuncCall |
  Func |
  Tuple |
  Name |
  Numeral |
  Text
);

export type FuncCall = {
  type: 'function-call';
  func: Expression;
  arg: Expression;
}

export type Func = {
  type: 'function';
  arg: Name;
  body: Expression;
};

export type Tuple = {
  type: 'tuple';
  members: Expression[];
};

export type Name = {
  type: 'name';
  value: string;
  token: TokenMap['name'];
};

export type Numeral = {
  type: 'number';
  value: number;
  token: TokenMap['number'];
};

export type Text = {
  type: 'text';
  value: string;
  token: TokenMap['text'];
};
