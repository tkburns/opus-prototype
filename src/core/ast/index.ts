import type { ProgramNode, Program } from './program.types';
import type { ExpressionNode } from './expression.types';
import type { PatternNode } from './pattern.types';
import type { BaseNode } from './base.types';

export * from './program.types';
export * from './pattern.types';
export * from './expression.types';
export * from './base.types';

export type AST = Program;
export type Node = (
  ProgramNode |
  ExpressionNode |
  PatternNode |
  BaseNode
);

