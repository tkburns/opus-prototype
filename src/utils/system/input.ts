import fs from 'fs';

export type Source = string;
export type Input = {
  content: string;
  source: Source;
};

export const Input = {
  fromString: (source: Source, content: string): Input =>
    ({ source, content }),

  fromFile: (file: string): Input => {
    const content = fs.readFileSync(file, 'utf-8');
    return { source: `file:${file}`, content };
  },

  fromStdin: (): Input => {
    const content = fs.readFileSync(process.stdin.fd, 'utf-8');
    return { source: 'stdin', content };
  }
};
