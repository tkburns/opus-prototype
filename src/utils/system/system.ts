import type { Prepend, Last } from '../tuple.types';
import type { UFn } from '../helper.types';


export interface Module<A, B> {
  run: (a: A) => B;
};

export interface System<A, B> {
  run: (a: A) => B;
}


type Modules<A, Ts extends unknown[]> = {
  [K in keyof Ts]: K extends keyof Prepend<Ts, A>
      ? Module<Prepend<Ts, A>[K], Ts[K]>
      : never;
};

type FirstArg<A> = [Module<A, unknown>, ...unknown[]];


export const system = <A, Ts extends unknown[]>(...modules: FirstArg<A> & Modules<A, Ts>): System<A, Last<Ts>> => {
  const castModules = modules as Module<A | Ts[number], Ts[number]>[];

  const run = castModules.reduce(
      (piped, m) => (a) => m.run(piped(a)),
      ((a: A | Ts[number]) => a)
  ) as UFn<A, Last<Ts>>;

  return {
    run
  };
};

