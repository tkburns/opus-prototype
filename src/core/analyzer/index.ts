import { Module } from '&/utils/system/system';
import type * as AST from '../ast';

// fuse analyze runs in future - just run sequentially for now

export const analyze: Module<AST.Program, AST.Program> = {
  run: (program) => {
    return program;
  }
};
