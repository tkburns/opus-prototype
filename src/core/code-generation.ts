import type * as AST from './ast.types';
import { Module } from '&/utils/system/system';
import { lines, code } from '&/utils/system/stringification';
import { transformByType } from '&/utils/system/tree-walker';
import { last } from '&/utils/list';


const program = (node: AST.Program) => lines(
  ...node.entries
    .map(node => node.type === 'declaration' ? declaration(node) : expression(node))
    .flatMap(code => [`${code};`, '']),
);


const declaration = (node: AST.Declaration) =>
  `const ${name(node.name)} = ${expression(node.expression)}`;

const expression = (node: AST.Expression): string => transformByType(node, {
  'function-application': funcApplication,
  'thunk-force': thunkForce,
  match,
  'function': func,
  thunk,
  tuple,
  name,
  atom,
  bool,
  number,
  text
});


const safeFnAppTypes = [
  'name',
  'function-application',
  'thunk-application',
  'match'
];

const funcApplication = (node: AST.FuncApplication) =>{
  const fn = expression(node.func);
  const arg = expression(node.arg);

  return safeFnAppTypes.includes(node.func.type)
    ? `${fn}(${arg})`
    : `(${fn})(${arg})`;
};

const thunkForce = (node: AST.ThunkForce) =>{
  const thunk = expression(node.thunk);

  return safeFnAppTypes.includes(node.thunk.type)
    ? `${thunk}()`
    : `(${thunk})()`;
};


type MatchOptions = { subject: string };
const match = (node: AST.Match, { subject = 'subject' }: Partial<MatchOptions> = {}) => {
  const exhaustive = last(node.clauses)?.pattern.type === 'wildcard-pattern';

  const clauses = node.clauses.map((clause, num) =>
    matchClause(clause, { subject, exhaustive, last: num === node.clauses.length - 1 })
  );

  let body;
  if (exhaustive) {
    body = clauses.join(' else ');
  } else {
    body = code`
      ${clauses.join(' else ')} else {
        throw new Error(\`no match for \${subject}\`);
      }
    `;
  }

  return code`
    ((${subject}) => {
      ${body}
    })(${expression(node.principal)})
  `;
};

type MatchClauseOptions = MatchOptions & { exhaustive: boolean; last: boolean }
const matchClause = (node: AST.MatchClause, options: MatchClauseOptions) => {
  if (options.exhaustive && options.last && node.pattern.type === 'wildcard-pattern') {
    return code`
      {
        return ${expression(node.body)};
      }
    `;
  } else {
    return code`
      if (${pattern(node.pattern, options)}) {
        return ${expression(node.body)};
      }
    `;
  }
};

const pattern = (node: AST.Pattern, options: MatchOptions): string => transformByType(node, [options], {
  'name-pattern': namePattern,
  'tuple-pattern': tuplePattern,
  'simple-literal-pattern': simpleLiteralPattern,
  'wildcard-pattern': wildcardPattern
});

const namePattern = (node: AST.NamePattern, { subject }: MatchOptions) =>
  `__opus_internals__.match.name(${subject}, ${name(node.name)})`;

const tuplePattern = (node: AST.TuplePattern, { subject }: MatchOptions) => {
  const memberMatches = node.members.map((member, index) => {
    const subjectMember = `${subject}._${index}`;
    return `() => ${pattern(member, { subject: subjectMember })}`;
  });

  return code`
    __opus_internals__.match.tuple(${subject}, [
      ${memberMatches.join(',\n')}
    ])
  `;
};

const simpleLiteralPattern = (node: AST.SimpleLiteralPattern, { subject }: MatchOptions) =>
  `__opus_internals__.match.simpleLiteral(${subject}, ${expression(node.value)})`;

const wildcardPattern = (node: AST.WildcardPattern) =>
  'true /* wildcard */';


const func = (node: AST.Func) => {
  return `(${name(node.arg)}) => ${funcBody(node.body)}`;
};

const funcBody = (body: AST.Expression) => {
  const isObject = body.type === 'tuple';
  return isObject
    ? `(${expression(body)})`
    : expression(body);
};

const thunk = (node: AST.Thunk) => {
  return `() => ${funcBody(node.body)}`;
};

const tuple = (node: AST.Tuple) => {
  const memberFields = node.members.map((member, index) =>
    `_${index}: ${expression(member)}`
  );

  return code`
    {
      __opus_kind__: 'tuple',
      size: ${node.members.length},
      ${memberFields.join(',\n')}
    }
  `;
};

const name = (node: AST.Name) => node.value;

const atom = (node: AST.Atom) => `Symbol.for('${node.value}')`;

const bool = (node: AST.Bool) => node.value.toString();

const number = (node: AST.Numeral) => node.token.source;

const text = (node: AST.Text) => `"${node.value}"`;


export const codeGenerator: Module<AST.Node, string> = {
  run: (node) => {
    if (node.type !== 'program') {
      throw new Error(`cannot generate code directly for a ${node.type}; only 'program' is supported`);
    }

    return program(node);
  }
};
