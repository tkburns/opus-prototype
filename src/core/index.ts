import { collect } from "&/utils/system/collect";
import { system } from "&/utils/system/system";
import { lexer, Token } from './lexer';

export const core = system(
  lexer,
  collect<Token>()
);
