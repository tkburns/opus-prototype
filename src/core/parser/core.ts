import { repeated, optional, choice, repeatedRequired } from '&/utils/system/parser';
import { lrec } from '&/utils/system/parser/lrec';
import { ConsumeHandle } from '&/utils/system/parser/handles';
import type * as Base from '&/utils/system/parser/common.types';
import { cached } from '&/utils/system/parser/cache';
import { precedented } from '&/utils/system/parser/precedence';

import { FilteredToken } from '../lexer';
import type * as AST from '../ast';

import { name, RDParser } from './base';
import { pattern } from './match';
import { particle } from './particles';


type TopPrecedenceParser<Node, C = object> =
  Base.ExtendedRDParser<ConsumeHandle<FilteredToken>, C, [precedence?: number], Node>;
type PrecedenceParser<Node, C = object> =
  Base.ExtendedRDParser<ConsumeHandle<FilteredToken>, C, [precedence: number], Node>;

export const declaration: RDParser<AST.Declaration> = cached((handle, ctx) => {
  const vrb = name(handle, ctx);
  handle.consume('=');
  const expr = expression(handle, ctx);

  return {
    type: 'declaration',
    name: vrb,
    expression: expr,
  };
});

const precedenceCacheKey = (ctx: object, precedence: number) => `prec=${precedence}`;
const topPrecedenceCacheKey = (ctx: object, precedence = 0) =>
  precedenceCacheKey(ctx, precedence);

export const expression: TopPrecedenceParser<AST.Expression> = lrec.cached(
  topPrecedenceCacheKey,
  (handle, ctx, precedence = 0) => {
    return precedented(handle, ctx, { precedence, rec: expression }, [
      [match, funcApplication],
      [thunkForce],
      [parens, literal, name]
    ]);
  }
);


export const blockExpression: RDParser<AST.BlockExpression | AST.Expression> = cached((handle, ctx) => {
  const [entries] = repeatedRequired(1, handle, ctx, () => {
    const entry = choice(handle, ctx, [
      // need to import
      declaration,
      expression
    ]);

    optional(handle, ctx, () => {
      handle.consume(';');
    });

    return entry;
  });

  if (entries.length === 1 && entries[0].type !== 'declaration') {
    return entries[0];
  }

  return {
    type: 'block-expression',
    entries
  };
});


export const parens: RDParser<AST.Expression> = cached((handle, ctx) => {
  handle.consume('(');
  const expr = blockExpression(handle, ctx);
  handle.consume(')');

  return expr;
});


export const funcApplication: PrecedenceParser<AST.FuncApplication> = cached(
  precedenceCacheKey,
  (handle, ctx, precedence) => {
    const func = expression(handle, ctx, precedence);
    const arg = expression(handle, ctx, precedence + 1);

    return {
      type: 'function-application',
      func,
      arg,
    };
  }
);

export const thunkForce: PrecedenceParser<AST.ThunkForce> = cached(
  precedenceCacheKey,
  (handle, ctx, precedence) => {
    handle.consume('!');
    const thunk = expression(handle, ctx, precedence);

    return {
      type: 'thunk-force',
      thunk,
    };
  }
);


export const match: PrecedenceParser<AST.Match> = cached(
  precedenceCacheKey,
  (handle, ctx, precedence) => {
    const principal = expression(handle, ctx, precedence);
    handle.consume('match');
    const clauses = matchClauses(handle, ctx);

    return {
      type: 'match',
      principal,
      clauses
    };
  }
);

const matchClauses: RDParser<AST.MatchClause[]> = (handle, ctx) => {
  handle.consume('(');

  const [clauses] = repeatedRequired(1, handle, ctx, () => {
    const clause = matchClause(handle, ctx);

    optional(handle, ctx, () => {
      handle.consume(';');
    });

    return clause;
  });

  handle.consume(')');

  return clauses;
};

const matchClause: RDParser<AST.MatchClause> = (handle, ctx) => {
  const pat = pattern(handle, ctx);
  handle.consume('=>');
  const body = expression(handle, ctx);

  return {
    type: 'match-clause',
    pattern: pat,
    body
  };
};


export const literal: RDParser<AST.Literal> = (handle, ctx) => {
  return choice(handle, ctx, [
    func,
    thunk,
    tuple,
    particle
  ]);
};


export const func: RDParser<AST.Func> = cached((handle, ctx) => {
  const arg = name(handle, ctx);
  handle.consume('=>');
  const expr = expression(handle, ctx);

  return {
    type: 'function',
    arg,
    body: expr,
  };
});

export const thunk: RDParser<AST.Thunk> = cached((handle, ctx) => {
  handle.consume('{');
  const body = blockExpression(handle, ctx);
  handle.consume('}');

  return {
    type: 'thunk',
    body,
  };
});

export const tuple: RDParser<AST.Tuple> = cached((handle, ctx) => {
  handle.consume('(');

  const [members] = repeated(handle, ctx, () => {
    const member = expression(handle, ctx);
    optional(handle, ctx, () => {
      handle.consume(',');
    });

    return member;
  });

  handle.consume(')');

  return {
    type: 'tuple',
    members,
  };
});
