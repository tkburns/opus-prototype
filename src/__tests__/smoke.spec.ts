import { core } from '../core';
import { Input } from '../utils/system/input';
import fs from 'fs';

jest.mock('&/core/runtime', () => {
  const mockRuntime = `
/* runtime goes here */
  `;

  return {
    __esModule: true,
    injectRuntime: {
      run: (output: string) => {
        return mockRuntime + '\n\n' + output;
      }
    }
  };
});


const source = Input.fromString('raw', `
(* comment *)
boolean = true;
number = 12;
atom = :hello;
text = "hello world";

tuple = (atom, text);

(* functions *)
pair = a => b => (a, b);
pair (pair 1 2) (pair 3 4);

(* inline function *)
(a => b => (b, a)) 1 2;

(* pattern match *)
message = name => name match (
  :hello => "hello world";
  :bye => "bye everyone";
  _ => "other";
);

(* ------------------- *)
(* precedence *)

a = :a; b = :b; c = :c; d = :d;
f = x => x; g = x => x;

x => x match (_ => x);
x => f x;

f g a;
f (g a);

(x => a, y => b);
(a match (_ => a), f a);

f b match (_ => a);

x => a match (_ => y => b) (z => c) (d, f x);
`);

describe('smoke test (snapshot)', () => {
  it('matches snapshot', () => {
    fs.readFileSync(require.resolve('&&/runtime/main'), 'utf-8');
    expect(core().run(source)).toMatchSnapshot();
  });
});
