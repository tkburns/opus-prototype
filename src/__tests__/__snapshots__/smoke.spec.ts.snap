// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smoke test (snapshot) matches snapshot 1`] = `
"
/* runtime goes here */


const boolean = true;

const number = 12;

const atom = Symbol.for('hello');

const text = 'hello world';

const tuple = {
  __opus_kind__: 'tuple',
  size: 2,
  _0: atom,
  _1: text,
};

const block = (() => {
  'ignored';
  const alsoIgnored = 'foo';
  return 'returned';
})();

const pair = (a) => (b) => ({
  __opus_kind__: 'tuple',
  size: 2,
  _0: a,
  _1: b,
});

pair(pair(1)(2))(pair(3)(4));

((a) => (b) => ({
  __opus_kind__: 'tuple',
  size: 2,
  _0: b,
  _1: a,
}))(1)(2);

const message = (name) => ((subject) => {
  if (__opus_internals__.match.simpleLiteral(subject, Symbol.for('hello'))) {
    return 'hello world';
  } else if (__opus_internals__.match.simpleLiteral(subject, Symbol.for('bye'))) {
    return 'bye everyone';
  } else {
    return 'other';
  }
})(name);

((subject) => {
  if (__opus_internals__.match.tuple(subject, [
    () => __opus_internals__.match.simpleLiteral(subject._0, 1),
    () => true /* wildcard */,
  ])) {
    return '1';
  } else if (__opus_internals__.match.tuple(subject, [
    () => true /* wildcard */,
    () => __opus_internals__.match.simpleLiteral(subject._1, 2),
  ])) {
    return '2';
  } else {
    throw new Error(\`no match for \${subject}\`);
  }
})({
  __opus_kind__: 'tuple',
  size: 2,
  _0: 1,
  _1: 2,
});

((subject) => {
  if (__opus_internals__.match.name(subject, true)) {
    return (() => {
      const ret = 'true';
      return ret;
    })();
  } else if (__opus_internals__.match.name(subject, false)) {
    return (() => {
      const ret = 'false';
      return ret;
    })();
  } else {
    throw new Error(\`no match for \${subject}\`);
  }
})(boolean);

const one = () => 1;

const two = () => {
  1;
  return 2;
};

const oneAndTwo = () => {
  const id = (x) => x;
  const apply = (f) => (x) => f(x);
  return apply(id)({
    __opus_kind__: 'tuple',
    size: 2,
    _0: 1,
    _1: 2,
  });
};

const delay = (x) => () => x;

{
  __opus_kind__: 'tuple',
  size: 2,
  _0: delay(1)(),
  _1: (() => 2)(),
};

const a = Symbol.for('a');

const b = Symbol.for('b');

const c = Symbol.for('c');

const d = Symbol.for('d');

const f = (x) => x;

const g = (x) => x;

(x) => ((subject) => {
  {
    return x;
  }
})(x);

(x) => f(x);

f(g)(a);

f(g(a));

{
  __opus_kind__: 'tuple',
  size: 2,
  _0: (x) => a,
  _1: (y) => b,
};

{
  __opus_kind__: 'tuple',
  size: 2,
  _0: ((subject) => {
    {
      return a;
    }
  })(a),
  _1: f(a),
};

((subject) => {
  {
    return a;
  }
})(f(b));

(x) => ((subject) => {
  {
    return (y) => b;
  }
})(a)((z) => c)({
  __opus_kind__: 'tuple',
  size: 2,
  _0: d,
  _1: f(x),
});
"
`;
